ssl/quic/quic_stream.c

```diff
@@ -0,0 +1,512 @@
+     *
+     * A logical byte in the stream is
+     *
+     *   - in the NEW state if it is in new_set;
+     *   - is in the ACKED state if it is in acked_set
+     *       (and may or may not have been culled);
+     *   - is in the IN_FLIGHT state otherwise.
+     *
+     * Invariant: No logical byte is ever in both new_set and acked_set.
+     */
+    UINT_SET        new_set, acked_set;
+
+    /*
+     * The current size of the stream is ring_buf.head_offset. If
+     * have_final_size is true, this is also the final size of the stream.
+     */
+    unsigned int    have_final_size     : 1;
+    unsigned int    sent_final_size     : 1;
+    unsigned int    acked_final_size    : 1;
+};
+
+static void qss_cull(OSSL_QSS *qss);
+
+OSSL_QSS *ossl_qss_new(size_t init_buf_size)
+{
+    OSSL_QSS *qss;
+
+    qss = OPENSSL_zalloc(sizeof(OSSL_QSS));
+    if (qss == NULL)
+        return NULL;
+
+    ring_buf_init(&qss->ring_buf);
+    if (!ring_buf_resize(&qss->ring_buf, init_buf_size)) {
+        ring_buf_destroy(&qss->ring_buf);
+        OPENSSL_free(qss);
+        return NULL;
+    }
+
+    ossl_uint_set_init(&qss->new_set);
+    ossl_uint_set_init(&qss->acked_set);
+    return qss;
+}
+
+void ossl_qss_free(OSSL_QSS *qss)
+{
+    if (qss == NULL)
+        return;
+
+    ossl_uint_set_destroy(&qss->new_set);
+    ossl_uint_set_destroy(&qss->acked_set);
+    ring_buf_destroy(&qss->ring_buf);
+    OPENSSL_free(qss);
+}
+
+int ossl_qss_get_stream_frame(OSSL_QSS *qss,
+                              size_t skip,
+                              OSSL_QUIC_FRAME_STREAM *hdr,
+                              OSSL_QTX_IOVEC *iov,
+                              size_t *num_iov)
+{
+    size_t num_iov_ = 0, src_len = 0, total_len = 0, i;
+    uint64_t max_len;
+    const unsigned char *src = NULL;
+    UINT_SET_ITEM *range = qss->new_set.head;
+
+    if (*num_iov < 2)
+        return 0;
+
+    for (i = 0; i < skip && range != NULL; ++i)
+        range = range->next;
+
+    if (range == NULL) {
+        /* No new bytes to send, but we might have a FIN */
+        if (!qss->have_final_size || qss->sent_final_size)
+            return 0;
+
+        hdr->offset = qss->ring_buf.head_offset;
+        hdr->len    = 0;
+        hdr->is_fin = 1;
+        *num_iov    = 0;
+        return 1;
+    }
+
+    /*
+     * We can only send a contiguous range of logical bytes in a single
+     * stream frame, so limit ourselves to the range of the first set entry.
+     *
+     * Set entries never have 'adjacent' entries so we don't have to worry
+     * about them here.
+     */
+    max_len = range->range.end - range->range.start + 1;
+
+    for (i = 0;; ++i) {
+        if (total_len >= max_len)
+            break;
+
+        if (!ring_buf_get_buf_at(&qss->ring_buf,
+                                 range->range.start + total_len,
+                                 &src, &src_len))
+            return 0;
+
+        if (src_len == 0)
+            break;
+
+        assert(i < 2);
```

Comment:
hmmmmm.  We don't put asserts in the library.

Reply:
Wait, what? I've been using these extensively. libssl looks full of them.

Reply:
> hmmmmm. We don't put asserts in the library.

Err...yes we do? Generally `ossl_assert` is preferred so that we do something sensible in a non-debug build. `OPENSSL_assert` should not be used in the library since that causes a crash even in a production build, which we should not do.

Reply:
We can use assert(). Why would not we? Of course, it should be used only at places where there is no way to return an error. Or to check invariants in debug builds that cannot be checked in production builds. Otherwise ossl_assert() should be used. We should never use OPENSSL_assert except for cases where we really want to check an invariant even in production build and where even returning an error could do more harm than terminating the process immediately - i.e., severe memory corruption detected.